### PR TITLE
fix: add webhook validation of redirect usage

### PR DESCRIPTION
**Pull request by Xunzhuo**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Webhook validation should prevent a HTTPRoute that specifies a redirect filter from specifying a backendRef

**Which issue(s) this PR fixes**:

Fixes: https://github.com/kubernetes-sigs/gateway-api/issues/1779

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by robscott** [comment]
Hey @Xunzhuo, thanks for working on this! We really need more information in the PR description to help reviewers understand what the goals are. For example, https://github.com/kubernetes-sigs/gateway-api/pull/1753 has good detail for each section.

As far as the content of the PR, I think we need to have some broader discussion about this before merging. This PR would effectively prevent a case where redirect loop as proposed by @rainest in https://github.com/kubernetes-sigs/gateway-api/issues/1185 is possible. As I see it, we essentially have two options:

1. Say that redirect loop protection is likely not viable and instead require that both a redirect and backendRef cannot be specified in the same route rule (that's this PR).
2. Update the spec to define how redirect loop protection would work, enabling a single route rule to express something like "redirect to HTTPS, if already there, forward to backend foo". 

My gut tells me that 2 would be pretty challenging and that 1 is the most viable solution at this point, but we need to be very clear that this is a change to the API Spec and include a release note + API Spec update if that's the direction we go.

Since this would be a significant change, I want to hold off until we can discuss this at a community meeting + get some feedback on the high level direction of this PR first.

/hold
/cc @youngnick @rainest @sunjayBhatia @howardjohn

**Comment by Xunzhuo** [comment]
Yes @robscott, I am also a bit concerned about this change, since this is a user-facing change. We definitely need more discussion before pushing this forward.

Take this as a draft PR, changes can make after the final approach is out.

**Review by sunjayBhatia — COMMENTED** [review]
I may have missed some discussion here from a community meeting, but in Contour we currently ignore backendrefs if you've set a redirect filter so option 1 Rob describes seems fine

we could also specify that infinite redirect protection is recommended but implementation-specific support (or extended if we can write up a nice conformance test)

File: apis/v1alpha2/validation/httproute_test.go [hunk 1]
```diff
@@ -816,19 +817,43 @@ func TestValidateHTTPRouteTypeMatchesField(t *testing.T) {
...
 			},
 		},
+		backendRefs: []gatewayv1a2.HTTPBackendRef{{
+			BackendRef: gatewayv1a2.BackendRef{
+				BackendObjectReference: gatewayv1a2.BackendObjectReference{
+					Name: gatewayv1a2.ObjectName("test"),
+					Port: ptrTo(gatewayv1b1.PortNumber(8080)),
+				},
+			},
+		}},
```
Comment (robscott) [inline apis/v1alpha2/validation/httproute_test.go:827]: It looks like this is repeated a lot, can you move the standard set of backendRefs to a var so there's not quite so much repetition here?

Comment (robscott) [reply apis/v1alpha2/validation/httproute_test.go:827]: Actually it looks like this isn't modified at all for any of the test cases, maybe just add this to the route populated in `t.Run` below?

File: apis/v1beta1/validation/httproute.go [hunk 2]
```diff
@@ -74,6 +74,33 @@ func ValidateHTTPRouteSpec(spec *gatewayv1b1.HTTPRouteSpec, path *field.Path) fi
...
+				if len(rule.BackendRefs) != 0 {
+					errs = append(errs, field.Invalid(path.Index(ri).Child("filters").Index(fi), filter.RequestRedirect, "cannot specify backendRefs when using redirect filter"))
+				}
+				break
+			}
+		}
+		for bi, backendRef := range rule.BackendRefs {
+			for fi, filter := range backendRef.Filters {
+				if filter.RequestRedirect != nil {
+					errs = append(errs, field.Invalid(path.Index(ri).Child("backendRefs").Index(bi).Child("filters").Index(fi), filter.RequestRedirect, "cannot specify backendRef when using redirect filter"))
```
Comment (robscott) [inline apis/v1beta1/validation/httproute.go:97]:
```suggestion
					errs = append(errs, field.Invalid(path.Index(ri).Child("backendRefs").Index(bi).Child("filters").Index(fi), filter.RequestRedirect, "cannot specify RequestRedirect filter within a backendRef"))
```

File: apis/v1beta1/validation/httproute.go [hunk 3]
```diff
@@ -74,6 +74,33 @@ func ValidateHTTPRouteSpec(spec *gatewayv1b1.HTTPRouteSpec, path *field.Path) fi
...
+
+// validateInvalidBackendWithRedirectFilter validates if an HTTPRoute that specifies a redirect filter and specifying a backendRef
+func validateInvalidBackendWithRedirectFilter(rules []gatewayv1b1.HTTPRouteRule, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+
+	for ri, rule := range rules {
+		for fi, filter := range rule.Filters {
+			if filter.RequestRedirect != nil {
+				if len(rule.BackendRefs) != 0 {
+					errs = append(errs, field.Invalid(path.Index(ri).Child("filters").Index(fi), filter.RequestRedirect, "cannot specify backendRefs when using redirect filter"))
```
Comment (robscott) [inline apis/v1beta1/validation/httproute.go:89]:
```suggestion
					errs = append(errs, field.Invalid(path.Index(ri).Child("filters").Index(fi), filter.RequestRedirect, "cannot specify backendRefs when using RequestRedirect filter"))
```

File: apis/v1beta1/validation/httproute.go [hunk 4]
```diff
@@ -338,10 +365,6 @@ func hasExactlyOnePrefixMatch(matches []gatewayv1b1.HTTPRouteMatch) bool {
 	if len(matches) != 1 || matches[0].Path == nil {
 		return false
 	}
-	pathMatchType := matches[0].Path.Type
-	if *pathMatchType != gatewayv1b1.PathMatchPathPrefix {
-		return false
-	}
 
-	return true
+	return *matches[0].Path.Type == gatewayv1b1.PathMatchPathPrefix
```
Comment (robscott) [inline apis/v1beta1/validation/httproute.go:369]: This feels not very safe and also not relevant to the rest of the PR, maybe revert this change?

File: apis/v1beta1/validation/httproute_test.go [hunk 5]
```diff
@@ -997,19 +999,43 @@ func TestValidateHTTPRouteTypeMatchesField(t *testing.T) {
 				Remove: []string{"remove"},
 			},
 		},
+		backendRefs: []gatewayv1b1.HTTPBackendRef{{
+			BackendRef: gatewayv1b1.BackendRef{
+				BackendObjectReference: gatewayv1b1.BackendObjectReference{
+					Name: gatewayv1b1.ObjectName("test"),
+					Port: ptrTo(gatewayv1b1.PortNumber(8080)),
+				},
+			},
```
Comment (robscott) [inline apis/v1beta1/validation/httproute_test.go:1008]: This also looks like it's repeated a lot, can we use a var for this?

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @Xunzhuo!
